File: src/Frontend/projects/core/src/lib/models/dto/seasons/policy-data.ts
export class PolicyData {
    public id: number;
    public discountPolicyId: number;
    public createTime: Date;
    public internetKeyPartnerImportanceDiscount: number;
    public networkPartnerImportanceDiscount: number;
    public keyPartnerImportanceDiscount: number;
    public internetPartnerImportanceDiscount: number;
    public wholesalePartnerImportanceDiscount: number;
    public newPartnerDiscount: number;

    public repeatedPartnerDiscount: number;
    public purchaseAndSaleDiscount: number;
    public preOrderDiscount: number;
    public commission: number;
    public freeWarehouseCurrentOrderDiscount: number;
    public freeWarehousePastOrderDiscount: number;
    public markupForMismatchOfVolume: number;
    public plannedInstallment: number;
    public plannedPrepayment: number;
    public annualRate: number;
    public maxCountOfInstallmentPeriods: number;
    // array of available fields
    public volumeDiscount: Array<{ value: number, discount: number }> = [];
    public brandMixDiscount: Array<{ value: number, discount: number }> = [];
    public prepaymentDiscount: Array<{ value: number, discount: number }> = [];

    static fromJSON(obj: any) {
        if (!obj) {
            return null;
        }
        return Object.assign(
            new PolicyData(),
            obj,
            {
                createTime: obj.createTime ? new Date(obj.createTime) : null,
                volumeDiscount: obj.volumeDiscount ? JSON.parse(obj.volumeDiscount) : null,
                brandMixDiscount: obj.brandMixDiscount ? JSON.parse(obj.brandMixDiscount) : null,
                prepaymentDiscount: obj.prepaymentDiscount ? JSON.parse(obj.prepaymentDiscount) : null,
            },
        );
    }
}

File: src/Backend/Common/Models/Seasons/PolicyData.cs
﻿using System;

namespace Company.Common.Models.Seasons
{
    public class PolicyData
    {
        public int Id { get; set; }

        public int DiscountPolicyId { get; set; }

        public DiscountPolicy DiscountPolicy { get; set; }

        public DateTime CreateTime { get; set; }

        public DateTime? DeleteTime { get; set; }

        public int CreatorId { get; set; }

        public int? DeleterId { get; set; }

        /// ===========================
        /// ==       Constants       ==
        /// ===========================

        /// <summary>
        /// %
        /// </summary>
        public float InternetKeyPartnerImportanceDiscount { get; set; }

        public float NetworkPartnerImportanceDiscount { get; set; }

        public float KeyPartnerImportanceDiscount { get; set; }

        public float InternetPartnerImportanceDiscount { get; set; }

        public float WholesalePartnerImportanceDiscount { get; set; }

        /// <summary>
        /// %
        /// </summary>
        public float NewPartnerDiscount { get; set; }

        public float RepeatedPartnerDiscount { get; set; }

        /// <summary>
        /// %
        /// </summary>
        public float PurchaseAndSaleDiscount { get; set; }

        /// <summary>
        /// %
        /// </summary>
        public float Commission { get; set; }

        /// <summary>
        /// %
        /// </summary>
        public float PreOrderDiscount { get; set; }

        public float FreeWarehouseCurrentOrderDiscount { get; set; }

        public float FreeWarehousePastOrderDiscount { get; set; }

        /// <summary>
        /// %
        /// </summary>
        public float MarkupForMismatchOfVolume { get; set; }

        /// <summary>
        /// ед, months
        /// </summary>
        public int PlannedInstallment { get; set; }

        /// <summary>
        /// %
        /// </summary>
        public float PlannedPrepayment { get; set; }

        /// <summary>
        /// %
        /// </summary>
        public float AnnualRate { get; set; }

        /// <summary>
        /// ед, periods
        /// </summary>
        public int MaxCountOfInstallmentPeriods { get; set; }

        /// ============================
        /// ==          Lists         ==
        /// ============================

        /// <summary>
        /// %
        /// </summary>
        public string VolumeDiscount { get; set; }

        // Brand-mix discount, string dictionary numberOfBrands:%
        public string BrandMixDiscount { get; set; }

        // Prepayment discount, string dictionary %:%
        public string PrepaymentDiscount { get; set; }
    }
}
